### PR TITLE
Removing polling output to get the BDX messages from the BDX transfer…

### PR DESCRIPTION
**Pull request by nivi-apple**
… session

### CODE REVIEW

File: src/app/clusters/ota-requestor/BDXDownloader.cpp [hunk 1]
```diff
@@ -102,10 +109,11 @@ CHIP_ERROR BDXDownloader::SetBDXParams(const chip::bdx::TransferSession::Transfe
 
     VerifyOrReturnError(mState == State::kIdle, CHIP_ERROR_INCORRECT_STATE);
 
+    mBdxTransfer.RegisterOutputEventCallback(this, OnBDXEventReceived);
```
Comment (nivi-apple) [inline src/app/clusters/ota-requestor/BDXDownloader.cpp:112]: while reviewing right now I realized that the callback should be set when we PrepareforTransfer or Initiate transfer from requestor so maybe add a param to those API's rather than have a seperate register/unregister because its kind of required to have a callback for BDX to work. It seems more error prone to have a separate register/unregister. i will change this. Please lmk wdyt?

File: examples/ota-provider-app/linux/main.cpp [hunk 2]
```diff
@@ -306,9 +299,7 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
                              "  -x, --ignoreQueryImage <ignore count>\n"
                              "        The number of times to ignore the QueryImage Command and not send a response.\n"
                              "  -y, --ignoreApplyUpdate <ignore count>\n"
-                             "        The number of times to ignore the ApplyUpdateRequest Command and not send a response.\n"
-                             "  -P, --pollInterval <time in milliseconds>\n"
-                             "        Poll interval for the BDX transfer \n" };
+                             "        The number of times to ignore the ApplyUpdateRequest Command and not send a response.\n" };
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/main.cpp:302]: This same info is also in the README. Please also remove from there.

Comment (nivi-apple) [reply examples/ota-provider-app/linux/main.cpp:302]: I think this is for showing the options on the commandline for the user using the command

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp [hunk 3]
```diff
@@ -358,6 +353,22 @@ void OTAProviderExample::HandleQueryImage(app::CommandHandler * commandObj, cons
             OTAImageHeader header;
             VerifyOrDie(ParseOTAHeader(parser, mOTAFilePath, header) == true);
             VerifyOrDie(sizeof(mSoftwareVersionString) > header.mSoftwareVersionString.size());
+            // Return Image not available if we already updated to the version that's available
```
Comment (bzbarsky-apple) [inline examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:356]: This does not obviously seem related to the polling changes.  Why is it here?

File: src/app/clusters/ota-requestor/DefaultOTARequestor.h [hunk 4]
```diff
@@ -147,6 +147,12 @@ class DefaultOTARequestor : public OTARequestorInterface, public BDXDownloader::
                 Reset();
             }
 
+            if (!sendFlags.Has(chip::Messaging::SendMessageFlags::kExpectResponse))
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/DefaultOTARequestor.h:150]: This also seems unrelated to polling per se and should probably be a separate PR.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm [hunk 5]
```diff
@@ -175,6 +178,39 @@ static void HandleBdxInitReceivedTimeoutExpired(chip::System::Layer * systemLaye
         static_cast<BdxOTASender *>(state)->ResetState();
     }
 
+    bool HasTransferTimedOut()
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm:181]:
```suggestion
    bool TransferHasTimedOut()
```

File: src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm [hunk 6]
```diff
@@ -227,6 +263,14 @@ CHIP_ERROR OnTransferSessionBegin(TransferSession::OutputEvent & event)
             mSystemLayer->CancelTimer(HandleBdxInitReceivedTimeoutExpired, this);
         }
 
+        // Start a timer for the BDX transfer timeout
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm:266]: Why do we need a timer here, apart from the existing exchange response timer?  Maybe I'm just misunderstanding the setup, but it seems like the exchange timer exactly implements the "we expected a response by some time, but it did not come" behavior....

File: src/protocols/bdx/BdxTransferSession.h [hunk 7]
```diff
@@ -146,39 +147,33 @@ class DLL_EXPORT TransferSession
...
      *
      *   See OutputEventType for all possible output event types.
      *
-     * @param event     Reference to an OutputEvent struct that will be filled out with any pending output data
-     * @param curTime   Current time
+     * @param event     Reference to an OutputEvent struct that will be filled out with
+     *                  the generated output event
      */
-    void PollOutput(OutputEvent & event, System::Clock::Timestamp curTime);
+    using OutputEventCallback = void (*)(void * context, OutputEvent & event);
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:158]: What is the "context" param here?  Should be documented.

Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:158]: OutputEventHandler might be a better name, fwiw.

File: src/protocols/bdx/BdxTransferSession.h [hunk 8]
```diff
@@ -146,39 +147,33 @@ class DLL_EXPORT TransferSession
...
      *
      *   A TransferSession object must be initialized with either StartTransfer() or WaitForTransfer().
      *
      * @param role      Inidcates whether this object will be sending or receiving data
      * @param initData  Data for initializing this object and for populating a TransferInit message
      *                  The role parameter will determine whether to populate a ReceiveInit or SendInit
-     * @param timeout   The amount of time to wait for a response before considering the transfer failed
+     * @param callback  OutputEventCallback that is to be registered with the transfer session and gets called when Output events
+     *                  are generated.
+     * @param context   Reference to the caller of this method
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:171]: Does it have to be the caller?  Or is this just opaque context that will be passed to `callback`?  Seems like the latter.  And this should:

1. Document that this is what `context` is used for: calling `callback`.
2. Document the point until which that can happen (i.e. when it's safe to free `context`).

File: src/protocols/bdx/BdxTransferSession.h [hunk 9]
```diff
@@ -189,13 +184,15 @@ class DLL_EXPORT TransferSession
      * @param role            Inidcates whether this object will be sending or receiving data
      * @param xferControlOpts Indicates all supported control modes. Used to respond to a TransferInit message
      * @param maxBlockSize    The max Block size that this object supports.
-     * @param timeout         The amount of time to wait for a response before considering the transfer failed
+     * @param callback        OutputEventCallback that is to be registered with the transfer session and gets called when Output
+     * events are generated.
+     * @param context         Reference to the caller of this method
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:189]: See comments above.

File: src/protocols/bdx/BdxTransferSession.h [hunk 10]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
      */
     CHIP_ERROR VerifyProposedMode(const BitFlags<TransferControlFlags> & proposed);
 
+    void SendStatusReportWithError(StatusReportData statusReportData);
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:345]: Document?  Though I guess this is private...

File: src/protocols/bdx/BdxTransferSession.h [hunk 11]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
     void PrepareStatusReport(StatusCode code);
     bool IsTransferLengthDefinite() const;
 
+    template <typename MessageType>
+    void PrepareOutgoingMessageEvent(MessageType messageType);
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:350]: Again, document?

File: src/protocols/bdx/BdxTransferSession.h [hunk 12]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
+
+    /**
+     * @brief
+     *   Register a callback with the transfer session. The callback will be called to notify the caller about any messages received
+     *   or messages that need to be send or errors that occurred internally.
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:355]:
```suggestion
     *   or messages that need to be sent or errors that occurred internally.
```

File: src/protocols/bdx/BdxTransferSession.h [hunk 13]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
+     *
+     * @param callback    OutputEventCallback that is to be registered with the transfer session and gets called when Output events
+     *                    are generated.
+     * @param context     Reference to the caller of this method
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:359]: See comments above.

File: src/protocols/bdx/BdxTransferSession.h [hunk 14]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
+     *
+     */
+    void UnregisterOutputEventCallback();
+
+    /**
+     * @brief
+     *   Sends the generated output event from the transfer session - BDX messages, status reports, etc to
+     *   the consumers of the messages who register the OutputEventCallback with it.
+     */
+    void SendOutputEvent(chip::bdx::TransferSession::OutputEvent & outputEvent);
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:375]: This feels more like DispatchOutputEvent.

File: src/protocols/bdx/BdxTransferSession.h [hunk 15]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
+
+    /**
+     * The context that is passed in along with the callback. When the output event is generated, the context is passed when the
+     * OutputEventCallback is called.
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:379]:
```suggestion
     * The context that is passed to the registered OutputEventCalllback.
```
(or OutputEventHandler, whatever the naming).

File: src/protocols/bdx/BdxTransferSession.h [hunk 16]
```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
...
+     */
+    void * mContext;
     OutputEventType mPendingOutput = OutputEventType::kNone;
-    TransferState mState           = TransferState::kUnitialized;
+
+    /**
+     * The output event callback that is registered with the transfer session and which gets called when any messages are received
+     * or messages that need to be send or errors that occurred internally.
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:386]:
```suggestion
     * The output event callback that is registered with the transfer session and which gets notified about OutputEvents when they happen.
```

File: src/protocols/bdx/BdxTransferSession.h [hunk 17]
```diff
@@ -365,14 +400,9 @@ class DLL_EXPORT TransferSession
     uint64_t mTransferLength       = 0; ///< 0 represents indefinite length
     uint16_t mTransferMaxBlockSize = 0;
 
-    // Used to store event data before it is emitted via PollOutput()
+    // Used to store event data send via the OutputEventCallback
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.h:403]:
```suggestion
    // Used to store event data to be sent via the OutputEventCallback
```

File: src/protocols/bdx/BdxTransferSession.cpp [hunk 18]
```diff
@@ -64,82 +54,50 @@ TransferSession::TransferSession()
...
-    // If there's no other pending output but an error occurred or was received, then continue to output the error.
-    // This ensures that when the TransferSession encounters an error and needs to send a StatusReport, both a kMsgToSend and a
-    // kInternalError output event will be emitted.
-    if (event.EventType == OutputEventType::kNone && mState == TransferState::kErrorState)
-    {
-        event = OutputEvent::StatusReportEvent(OutputEventType::kInternalError, mStatusReportData);
-    }
+    OutputEvent event = TransferSession::OutputEvent::MsgToSendEvent(outputMsgType, std::move(mPendingMsgHandle));
+    SendOutputEvent(event);
+    mPendingMsgHandle = nullptr;
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.cpp:77]: Other places don't set it null like this.  Why do we need to do it here?

File: src/protocols/bdx/TransferFacilitator.cpp [hunk 19]
```diff
@@ -42,18 +37,19 @@ CHIP_ERROR TransferFacilitator::OnMessageReceived(chip::Messaging::ExchangeConte
...
 
-    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
-    // transfer could necessitate a response if they are received at the wrong time.
-    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
-    // transfer is finished.
-    mExchangeCtx->WillSendMessage();
+    // Almost every BDX message except BlockAckEOF will follow up with a response on the exchange. Even messages that might signify
+    // the end of a transfer could necessitate a response if they are received at the wrong time. For this reason, it is left up to
+    // the application logic to call ExchangeContext::Close() when it has determined that the transfer is finished.
+    if (mExchangeCtx != nullptr && !payloadHeader.HasMessageType(chip::bdx::MessageType::BlockAckEOF))
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/TransferFacilitator.cpp:49]: So do we send responses to StatusReport messages?

That said, if BlockAckEOF happens at the "wrong" time (e.g. when a block request is expected!), would _that_ not get a StatusResponse?

File: src/protocols/bdx/TransferFacilitator.cpp [hunk 20]
```diff
@@ -65,70 +61,30 @@ void TransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
...
-    HandleTransferSessionOutput(outEvent);
-
-    VerifyOrReturn(mSystemLayer != nullptr, ChipLogError(BDX, "%s mSystemLayer is null", __FUNCTION__));
-    if (!mStopPolling)
-    {
-        mSystemLayer->StartTimer(mPollFreq, PollTimerHandler, this);
-    }
-    else
+    ChipLogProgress(BDX, "OnOutputEventReceived %s", event.ToString(event.EventType));
+    TransferFacilitator * facilitator = static_cast<TransferFacilitator *>(context);
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/TransferFacilitator.cpp:67]:
```suggestion
    auto * facilitator = static_cast<TransferFacilitator *>(context);
```

File: src/protocols/bdx/TransferFacilitator.h [hunk 21]
```diff
@@ -34,19 +34,21 @@ namespace chip {
 namespace bdx {
 
 /**
- * An abstract class with methods for handling BDX messages from an ExchangeContext and polling a TransferSession state machine.
+ * An abstract class with methods for handling BDX messages from an ExchangeContext and gets output events from a TransferSession
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/TransferFacilitator.h:37]: This sentence has weird grammar...

File: src/protocols/bdx/TransferFacilitator.h [hunk 22]
```diff
@@ -107,18 +90,13 @@ class Responder : public TransferFacilitator
 {
 public:
     /**
-     * Initialize the TransferSession state machine to be ready for an incoming transfer request, and start the polling timer.
-     *
-     * @param[in] layer           A System::Layer pointer to use to start the polling timer
+     * Initialize the TransferSession state machine to be ready for an incoming transfer request and set the callback to get the
+     * output event
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/TransferFacilitator.h:94]: Is the "callback to get the output event" something the consumer of this API needs to know about?

File: src/protocols/bdx/TransferFacilitator.h [hunk 23]
```diff
@@ -133,18 +111,12 @@ class Initiator : public TransferFacilitator
 {
 public:
     /**
-     * Initialize the TransferSession state machine to prepare a transfer request message (does not send the message) and start the
-     * poll timer.
+     * Initialize the TransferSession state machine to prepare a transfer request message and sends the outgoing message
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/TransferFacilitator.h:114]: "prepare" but "sends"?  Hard to make sense of this sentence.

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp (hunk 3 above)
Comment (nivi-apple) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:356]: This is a bug in the OTA example code. when we test with 3 ota requestors, the first ota requestor successfully updates. we do a query image again on ota 1 and the version is already updated but still with old code we get image available and the ota provider prepares for transfer. now when ota 2 sends a query image it will get busy which is incorrect. The ios OTA provider already does that. we just add that behavior to the ota provider example

Comment (bzbarsky-apple) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:356]: I am not saying it's not a bug.  I am saying:

1. This chamge has nothing to do with the polling changes.
2. Therefore it should not be in this PR.
3. Moving this part to a separate PR will improve the odds that it will get careful review and make this already-large PR smaller and easier to review.

Comment (nivi-apple) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:356]: sounds good. will do that

File: src/app/clusters/ota-requestor/DefaultOTARequestor.h (hunk 4 above)
Comment (nivi-apple) [reply src/app/clusters/ota-requestor/DefaultOTARequestor.h:150]: I think this causes a crash on the requestor side when testing with the app and my code changes. Its been a while I tested this. So i am going to try again. I will update the PR comment with details.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm (hunk 6 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm:266]: With the poll output code the bdx timeout timer was in the transfer session itself. Last I spoke to Carol the idea was for the requestor and provider to track if BDX has stalled for more than 5 minutes and if it is we abort the transfer. The change was made for the requestor as part of the clean up but the provider was not updated. We discussed and made this change accordingly.

File: src/protocols/bdx/BdxTransferSession.h (hunk 8 above)
Comment (nivi-apple) [reply src/protocols/bdx/BdxTransferSession.h:171]: opaque context that will be passed to callback. will document

File: src/protocols/bdx/TransferFacilitator.cpp (hunk 19 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.cpp:49]: we do not send responses to status report messages. if a blockackeof happens at the wrong time a status report will be sent with appropriate error code. see - https://github.com/project-chip/connectedhomeip/pull/26040/commits/e5213ad18a9342187ec13de8795d643e18197a94#diff-eb27c4ea98e4f9214d05fd68debdcf2291a46ae3d74039b7b7e412aa55ee7708L728

File: src/protocols/bdx/TransferFacilitator.h (hunk 22 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.h:94]: no this was a find replace of the polling timer comments. will remove

File: src/protocols/bdx/TransferFacilitator.h (hunk 23 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.h:114]: same as above. not required

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp (hunk 3 above)
Comment (nivi-apple) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp]: done

File: src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm (hunk 5 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm]: done.

File: src/protocols/bdx/BdxTransferSession.h (hunk 7 above)
Comment (nivi-apple) [reply src/protocols/bdx/BdxTransferSession.h]: done.

File: src/protocols/bdx/TransferFacilitator.cpp (hunk 20 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.cpp]: done

File: src/protocols/bdx/TransferFacilitator.h (hunk 21 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.h]: done

File: src/protocols/bdx/TransferFacilitator.h (hunk 22 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.h]: done

File: src/protocols/bdx/TransferFacilitator.h (hunk 23 above)
Comment (nivi-apple) [reply src/protocols/bdx/TransferFacilitator.h]: done

File: src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm (hunk 6 above)
Comment (tcarmelveilleux) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm:266]: The time-out should be at the exchange layer if at all possible (see src/messaging/ExchangeContext.h):

```
// Set the response timeout for the exchange context, regardless of the underlying session type. Using
    // UseSuggestedResponseTimeout to set a timeout based on the type of the session and the application processing time instead of
    // using this function is recommended.
    //
    // If a timeout of 0 is provided, it implies no response is expected. Consequently, ExchangeDelegate::OnResponseTimeout will not
    // be called.
    //
    void SetResponseTimeout(Timeout timeout);
```

File: src/protocols/bdx/tests/TestBdxTransferSession.cpp [hunk 24]
```diff
@@ -113,183 +154,490 @@ void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::P
...
+    err = initiator.StartTransfer(initiatorRole, initData, OnInitiatorOutputEventReceived, inSuite);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Now process output events until we get to one that needs no action from
+    // us.
+    bool transferInProgress = true;
+    do
+    {
+        transferInProgress = RespondToLastInitiatorOutputEvent(inSuite);
+    } while (transferInProgress);
```
Comment (tcarmelveilleux) [inline src/protocols/bdx/tests/TestBdxTransferSession.cpp:533]: This entire unit test now contains practically more logic than the protocol itself, and could now contain bugs. It would have been great to avoid adding so much actual logic to the unit test.

File: src/protocols/bdx/tests/TestBdxTransferSession.cpp [hunk 25]
```diff
@@ -113,183 +154,490 @@ void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::P
...
+    blockData.Length = fakeDataLen;
+    blockData.IsEof  = false;
+
+    sInitiatorExpectedOutputEvent.msgTypeData.MessageType =
+        static_cast<uint8_t>(isEof ? MessageType::BlockEOF : MessageType::Block);
+    // Set the expected BDX event type
+    if (inBlockCounter == 0)
+    {
+        sResponderExpectedOutputEvent.EventType = TransferSession::OutputEventType::kBlockReceived;
+        sResponderExpectedOutputEvent.blockdata = blockData;
```
Comment (tcarmelveilleux) [inline src/protocols/bdx/tests/TestBdxTransferSession.cpp:626]: The blockData will be used after scope (since it is **non-owning** and pointing to a stack buffer). Please fix.

File: src/protocols/bdx/TransferFacilitator.cpp [hunk 26]
```diff
@@ -65,70 +61,30 @@ void TransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
...
-
-    mPollFreq    = pollFreq;
-    mSystemLayer = layer;
-
-    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
-
-    ChipLogProgress(BDX, "Start polling for messages");
-    mStopPolling = false;
-    mSystemLayer->StartTimer(mPollFreq, PollTimerHandler, this);
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, OnOutputEventReceived, this));
```
Comment (tcarmelveilleux) [inline src/protocols/bdx/TransferFacilitator.cpp:76]: What unblocks this if no events ever come?

File: src/protocols/bdx/BdxTransferSession.h [hunk 27]
```diff
@@ -365,14 +411,10 @@ class DLL_EXPORT TransferSession
...
 
-    // Used to store event data before it is emitted via PollOutput()
-    System::PacketBufferHandle mPendingMsgHandle;
-    StatusReportData mStatusReportData;
-    TransferInitData mTransferRequestData;
-    TransferAcceptData mTransferAcceptData;
-    BlockData mBlockEventData;
-    MessageTypeData mMsgTypeData;
-    TransferSkipData mBytesToSkip;
+    const uint8_t * mFileDesignator       = nullptr;
```
Comment (tcarmelveilleux) [inline src/protocols/bdx/BdxTransferSession.h:414]: This is non-owning. It points to memory possibly released. Why is this a non-owning pointer?

File: src/protocols/bdx/BdxTransferSession.cpp [hunk 28]
```diff
@@ -64,82 +59,50 @@ TransferSession::TransferSession()
     mSuppportedXferOpts.ClearAll();
 }
 
-void TransferSession::PollOutput(OutputEvent & event, System::Clock::Timestamp curTime)
+void TransferSession::DispatchOutputEvent(TransferSession::OutputEvent & outputEvent)
```
Comment (tcarmelveilleux) [inline src/protocols/bdx/BdxTransferSession.cpp:62]: This significantly changes the public API of BDX transfer session. PollOutput doesn't mean that it actually polls. It is a misnomer. PollOutput just means "give me the next action to take". Triggering that from messages is what was needed. This change of public API completely changes the directionality of the transfer session API. This will break external clients that were not using the TransferFacilitator.

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [ ] These changes to public  BdxTransferSession API break our (Google's) own usage of this code. If TransferFacilitator was the only one touched and not  BdxTransferSession, then we (but maybe not others) would be OK. 

Such public API changes must come with enough notice to make sure the cost of dealing with the change can be avoided if possible.

**Comment by nivi-apple** [comment]
Replaced by https://github.com/project-chip/connectedhomeip/pull/29364
